src/go/k8s/controllers/redpanda/cluster_controller.go

```diff
@@ -333,6 +351,146 @@ func (r *ClusterReconciler) setLicense(
+				//nolint:goerr113 // not going to use wrapped static error here this time
+				return fmt.Errorf(`unable to set the finalizer on pod "%s": %d`, pod.Name, err)
+			}
+			continue
+		}
+		// if the pod is being deleted
+		// check the node it's assigned to
+		node := corev1.Node{}
+		key := types.NamespacedName{Name: pod.Spec.NodeName}
+		err := r.Get(ctx, key, &node)
+		// if the node is not gone
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf(`unable to fetch node "%s": %w`, pod.Spec.NodeName, err)
+		}
+		if err == nil {
+			// nor has a noexecute taint
+			untainted := true
+			for _, taint := range node.Spec.Taints {
+				if taint.Effect == corev1.TaintEffectNoExecute && taint.Key == corev1.TaintNodeUnreachable {
+					untainted = false
+				}
+			}
+			if untainted {
+				// remove the finalizer and let the pod be restarted
+				if err = r.removePodFinalizer(ctx, pod, log); err != nil {
+					return fmt.Errorf(`unable to remove finalizer from pod "%s": %w`, pod.Name, err)
+				}
+				continue
+			}
+		}
+		// get the node id
+		nodeIDStr, ok := pod.GetAnnotations()[PodAnnotationNodeIDKey]
+		if !ok {
+			return fmt.Errorf("cannot determine node_id for pod %s: %w. not removing finalizer", pod.Name, err)
+		}
+		nodeID, err := strconv.Atoi(nodeIDStr)
+		if err != nil {
+			return fmt.Errorf("node-id annotation is not an integer: %w", err)
+		}
+		// get the broker list
+		redpandaPorts := networking.NewRedpandaPorts(rp)
+		headlessPorts := collectHeadlessPorts(redpandaPorts)
+		clusterPorts := collectClusterPorts(redpandaPorts, rp)
+		headlessSvc := resources.NewHeadlessService(r.Client, rp, r.Scheme, headlessPorts, log)
+		clusterSvc := resources.NewClusterService(r.Client, rp, r.Scheme, clusterPorts, log)
+
+		pki := certmanager.NewPki(r.Client, rp, headlessSvc.HeadlessServiceFQDN(r.clusterDomain), clusterSvc.ServiceFQDN(r.clusterDomain), r.Scheme, log)
+
+		adminClient, err := r.AdminAPIClientFactory(ctx, r.Client, rp, headlessSvc.HeadlessServiceFQDN(r.clusterDomain), pki.AdminAPIConfigProvider())
+		if err != nil {
+			return fmt.Errorf("unable to create admin client: %w", err)
+		}
+		brokers, err := adminClient.Brokers(ctx)
+		if err != nil {
+			return fmt.Errorf("unable to fetch brokers: %w", err)
+		}
+		// check if the node in the broker list
+		var broker *admin.Broker
+		for i := range brokers {
+			if brokers[i].NodeID == nodeID {
+				broker = &brokers[i]
+				break
+			}
+		}
+		// if it's not gone
+		if broker != nil {
+			// decommission it
+			log.WithValues(nodeID).Info("decommissioning broker")
+			if err = adminClient.DecommissionBroker(ctx, nodeID); err != nil {
+				return fmt.Errorf(`unable to decommission node "%d": %w`, nodeID, err)
+			}
+		}
+		//   delete the associated pvc
+		pvc := corev1.PersistentVolumeClaim{}
+		//nolint: gocritic // 248 bytes 6 times is not worth decreasing the readability over
+		for _, v := range pod.Spec.Volumes {
```

Comment:
This deletes all PVCs attached to our pods regardless of its nature.
In some cases ( eg EBS/network-backed PVCs ) it might make sense to keep those, although it wholly depends on the model we want to follow.
This probably makes the operator's behavior differ from previous version (probably backwards incompatible wrt data handling)

Would it be worth to have a config so we delete only if the pvc affinity is on the node and keep things that are bound to eg an Availability Zone?

Reply:
This is now solved by adding a command-line flag to enable this behavior (disabled by default).